Configure alertmanager

wenxin-liu opened the pull request.

wenxin-liu committed:
Added alertmanager as a plugin to be installed

wenxin-liu committed:
Testing creating an alertmanager datasource with a configuration file

wenxin-liu committed:
fixed error

wenxin-liu committed:
fixed error in alertmanager datasource configuration yaml

wenxin-liu committed:
Cosmetic change

wenxin-liu committed:
Changed the URL of the alert manager datasource

wenxin-liu committed:
test

wenxin-liu committed:
test

wenxin-liu committed:
test

wenxin-liu committed:
Revert URL of alertmanager datasource

wenxin-liu committed:
test

wenxin-liu committed:
revert testing

wenxin-liu committed:
Configured the alertmanager datasource URL to be the right one hopefully

wenxin-liu committed:
Right URL for alertmanager datasource

antoniocascais commented on grafana.sh:
```diff
@@ -279,14 +279,59 @@ set_vcap_datasource_prometheus() {
     [[ -z "${datasource}" ]] && datasource=$(get_prometheus_vcap_service)
     if [[ -n "${datasource}" ]]
     then
+        echo "camptocamp-prometheus-alertmanager-datasource 0.0.7" > ${GRAFANA_CFG_PLUGINS}
```
We should use `>>` here, to avoid overriding the whole plugins file

wenxin-liu replied:
This is a good point. Should be `>>` and not `>`.

Separately, through speaking to Jose, he proposed a whole variety of changes across various files (instead of hard-coding 0.0.7). I didn't agree with 100% of his proposals, but some of it I agreed with.

So this is where the bulk of the refactoring work is taking place.

I'll make a new commit when I've refactored.

wenxin-liu replied:
I've refactored and committed new changes regarding the hard-coded version number.

antoniocascais commented on grafana.sh:
```diff
@@ -263,7 +263,7 @@ set_vcap_datasource_prometheus() {
 	# what's available in the database
 	datasources:
-	- name: ${name}
+	- name: ${name} Prometheus
```
Is this correct?

wenxin-liu replied:
Yes, this is right. 

Before, the name of the Prometheus datasource was just the name of the service instance, e.g.:
wenxin-new-new-prometheus-test.

Now that we have multiple datasources, I want to make the names more specific, and differentiate Prometheus from Alertmanager.

So now there'll be two datasources, called:
wenxin-new-new-prometheus-test Prometheus
wenxin-new-new-prometheus-test AlertManager

antoniocascais commented on grafana.sh:
```diff
@@ -279,14 +279,59 @@ set_vcap_datasource_prometheus() {
+    if [[ ${datasources_folder_exists_exit_code} -eq 1 ]]
+    then
+        # Only create datasources directory if it is not already created in set_vcap_datasource_prometheus()
+        mkdir -p "${APP_ROOT}/datasources"
```
We can execute this command even if the directory exists. It won't do anything if the directory already exists. Unless there's some logic I'm missing...

wenxin-liu replied:
When I test it locally, I get an error when the directory already exists:
Wenxins-MacBook-Pro:gcp-service-broker wenxinliu1$ mkdir dc/test
mkdir: dc/test: File exists

At the top of the script, `set -e` is applied. Jose wrote `set -euo pipefail`. So I don't want the script to exit on error.

Equally I could simplify and just write:
        `set +e`
        `mkdir -p "${APP_ROOT}/datasources"`
       `set -e`
instead of setting variables and checking with the if condition. And allow the script to continue, even if it's erroring.

Here, I don't think it makes much of a difference, I feel like it's cleaner there's some way to handle the error, rather than just erroring and continuing.

So I think I'll keep this section as is.

antoniocascais replied:
```
 $ mkdir -p /tmp/stuff
 $ echo $?
0
$ mkdir -p /tmp/stuff
$ echo $?
0
```
As you can see, the exit code is always 0. 

```
Wenxins-MacBook-Pro:gcp-service-broker wenxinliu1$ mkdir dc/test
mkdir: dc/test: File exists
```
You forgot the `-p` flag in your test :)

wenxin-liu replied:
Yes, I didn't see the `-p` flag/know about the change to functionality `-p` provides! Thank you, good point! I see the point you're making now about the unnecessary logic in this comment.

I now deleted the unnecessary logic, leaving only the `mkdir -p` line. I made a new commit that I've now pushed to this branch.

wenxin-liu commented:
Thank you Antonio for your review and suggestions!

wenxin-liu committed:
Extrapolated AlertManager version number into a variable whose value can be overwritten in CF manifest

The user can define their own AlertManager version number in their CF manifest, in the env block, using variable name GRAFANA_ALERTMANAGER_VERSION. Otherwise version number defaults to 0.0.7, as specified in parameters.sh. This is so no values are hard-coded into grafana.sh, and future changes to the default value of AlertManager can be done in parameters.sh, where other version numbers are also stored.

wenxin-liu committed:
Removed unnecessary logic
